File: beam-android/core/testing/src/main/kotlin/dev/beam/android/core/testing/FakeRepositories.kt
package dev.beam.android.core.testing

import dev.beam.android.core.ffi.preferences.PreferencesRepository
import dev.beam.android.core.ffi.repository.AdminRepository
import dev.beam.android.core.ffi.repository.CatalogRepository
import dev.beam.android.core.ffi.repository.PlaybackRepository
import dev.beam.android.core.ffi.repository.ServerRepository
import dev.beam.android.core.ffi.repository.SessionRepository
import dev.beam.android.core.model.UserPreferences
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import uniffi.beam_client_core.AdminEvent
import uniffi.beam_client_core.AdminLogEntry
import uniffi.beam_client_core.AdminStatus
import uniffi.beam_client_core.AdminUser
import uniffi.beam_client_core.AdminUserPage
import uniffi.beam_client_core.BeamException
import uniffi.beam_client_core.BrowseQuery
import uniffi.beam_client_core.ContinueWatchingEntry
import uniffi.beam_client_core.DeviceProfile
import uniffi.beam_client_core.DeviceSession
import uniffi.beam_client_core.EpisodeSummary
import uniffi.beam_client_core.HistoryPage
import uniffi.beam_client_core.LibraryFileSummary
import uniffi.beam_client_core.LibrarySummary
import uniffi.beam_client_core.MediaDetail
import uniffi.beam_client_core.MediaPage
import uniffi.beam_client_core.MediaSourceView
import uniffi.beam_client_core.PlaybackHttpConfig
import uniffi.beam_client_core.ProgressOutcome
import uniffi.beam_client_core.QualityPolicy
import uniffi.beam_client_core.ServerHealth
import uniffi.beam_client_core.ServerSummary
import uniffi.beam_client_core.SessionState
import uniffi.beam_client_core.SourceSelection
import uniffi.beam_client_core.UserSummary

// Stateful fakes rather than mocks, per the repository's testing rules: a
// screen that adds a server and then lists servers should see it, and only a
// fake that actually holds state can prove that. Each carries a `failWith`
// so the error paths the repo requires to be codified are one line to set up.

/** A catalog that answers from memory. */
public class FakeCatalogRepository : CatalogRepository {
    /** Set to make every call throw, for testing failure rendering. */
    public var failWith: BeamException? = null

    /** Pages returned by [browse], in call order; the last repeats. */
    public var pages: List<MediaPage> = listOf(Fixtures.page())

    /** Details keyed by media id. */
    public var details: MutableMap<String, MediaDetail> = mutableMapOf()

    /** Genres returned by [genres]. */
    public var genreList: List<String> = listOf("Crime", "Drama", "Thriller")

    /** Libraries returned by [libraries]. */
    public var libraryList: List<LibrarySummary> = listOf(Fixtures.library())

    /** Files returned by [libraryFiles]. */
    public var files: List<LibraryFileSummary> = emptyList()

    /** The episode [upNext] resolves to. */
    public var nextEpisode: EpisodeSummary? = null

    /** Every query [browse] was called with, in order. */
    public val browseCalls: MutableList<BrowseQuery> = mutableListOf()

    /**
     * The cursor every page reports, or null for "this is the last page".
     *
     * Overrides whatever [pages] carries, so a paging test can be written
     * without constructing a page fixture per page.
     */
    public var nextCursor: String? = null

    /**
     * Set to make [genres] alone fail.
     *
     * Distinct from [failWith] because the genre chips and the catalog grid
     * are separate concerns: one failing must not be indistinguishable from
     * both failing.
     */
    public var genresFailWith: BeamException? = null

    override suspend fun browse(query: BrowseQuery): MediaPage {
        failWith?.let { throw it }
        browseCalls += query
        val page = pages.getOrElse(browseCalls.size - 1) { pages.last() }
        return page.copy(endCursor = nextCursor, hasNextPage = nextCursor != null)
    }

    override suspend fun detail(mediaId: String): MediaDetail {
        failWith?.let { throw it }
        return details[mediaId]
            ?: throw BeamException.NotFound("No title with id $mediaId")
    }

    override suspend fun genres(): List<String> {
        genresFailWith?.let { throw it }
        failWith?.let { throw it }
        return genreList
    }

    override suspend fun libraries(): List<LibrarySummary> {
        failWith?.let { throw it }
        return libraryList
    }

    override suspend fun library(libraryId: String): LibrarySummary {
        failWith?.let { throw it }
        return libraryList.firstOrNull { it.id == libraryId }
            ?: throw BeamException.NotFound("No library with id $libraryId")
    }

    override suspend fun libraryFiles(libraryId: String): List<LibraryFileSummary> {
        failWith?.let { throw it }
        return files
    }

    override suspend fun upNext(
        showId: String,
        currentEpisodeId: String,
    ): EpisodeSummary? {
        failWith?.let { throw it }
        return nextEpisode
    }
}

/** A playback surface that records what it was told. */
public class FakePlaybackRepository : PlaybackRepository {
    /** Set to make every call throw. */
    public var failWith: BeamException? = null

    /** The profile last handed to [setDeviceProfile]. */
    public var deviceProfile: DeviceProfile? = null

    /** Sources returned by [sources]. */
    public var sourceList: List<MediaSourceView> = listOf(Fixtures.source())

    /** The selection returned by [selectSource]. */
    public var selection: SourceSelection? = null

    /** Rows returned by [continueWatching]. */
    public var continueWatchingRows: List<ContinueWatchingEntry> =
        listOf(Fixtures.continueWatching())

    /** The page returned by [history]. */
    public var historyPage: HistoryPage = HistoryPage(listOf(Fixtures.historyEntry()), 1uL)

    /**
     * Every progress report, in order.
     *
     * Records the duration and the force flag as well as the position, because
     * the interesting assertions are about *those*: whether a pause forced a
     * send, and whether an unknown duration was sent as absent rather than as
     * a confident zero.
     */
    public val reportedProgress: MutableList<ProgressReport> = mutableListOf()

    /** What [reportProgress] returns. */
    public var progressOutcome: ProgressOutcome = ProgressOutcome.Sent(0.0)

    /** How many entries [flushProgress] claims to have sent. */
    public var flushed: UInt = 0u

    /** How many entries are waiting. */
    public var pending: UInt = 0u

    override suspend fun setDeviceProfile(profile: DeviceProfile) {
        deviceProfile = profile
    }

    override suspend fun sources(mediaId: String): List<MediaSourceView> {
        failWith?.let { throw it }
        return sourceList
    }

    /** The policy [selectSource] was last asked for. */
    public var selectionPolicy: QualityPolicy? = null
        private set

    override suspend fun selectSource(
        mediaId: String,
        policy: QualityPolicy,
    ): SourceSelection {
        selectionPolicy = policy
        failWith?.let { throw it }
        return selection ?: throw BeamException.NotFound("This title has no playable files")
    }

    override suspend fun playbackConfig(fileId: String): PlaybackHttpConfig {
        failWith?.let { throw it }
        return PlaybackHttpConfig(
            url = "https://beam.test/v1/files/$fileId/stream",
            headers = mapOf("Cookie" to "beam_session=test"),
            trustedFingerprints = emptyList(),
            pinnedHost = "beam.test",
        )
    }

    override suspend fun continueWatching(limit: UInt?): List<ContinueWatchingEntry> {
        failWith?.let { throw it }
        return continueWatchingRows
    }

    override suspend fun history(
        limit: UInt?,
        offset: UInt?,
    ): HistoryPage {
        failWith?.let { throw it }
        return historyPage
    }

    override suspend fun reportProgress(
        fileId: String,
        positionSecs: Double,
        durationSecs: Double?,
        force: Boolean,
    ): ProgressOutcome {
        reportedProgress += ProgressReport(fileId, positionSecs, durationSecs, force)
        failWith?.let { throw it }
        return progressOutcome
    }

    override suspend fun flushProgress(): UInt {
        failWith?.let { throw it }
        return flushed
    }

    override suspend fun pendingProgressCount(): UInt = pending
}

/** A server registry that actually holds servers. */
public class FakeServerRepository(
    initial: List<ServerSummary> = listOf(Fixtures.server()),
) : ServerRepository {
    /** Set to make every call throw. */
    public var failWith: BeamException? = null

    private val state = MutableStateFlow(initial)

    override val servers: Flow<List<ServerSummary>> = state

    /** The cookie last handed to [completeLogin]. */
    public var capturedCookie: String? = null

    override suspend fun restore(): List<ServerSummary> {
        failWith?.let { throw it }
        return state.value
    }

    override suspend fun addServer(
        baseUrl: String,
        displayName: String?,
    ): ServerSummary {
        failWith?.let { throw it }
        val added =
            Fixtures.server(
                id = baseUrl.filter { it.isLetterOrDigit() },
                displayName = displayName ?: baseUrl,
                state = SessionState.LoggedOut,
            )
        state.value = state.value.map { it.copy(isActive = false) } + added
        return added
    }

    override suspend fun selectServer(serverId: String) {
        failWith?.let { throw it }
        state.value = state.value.map { it.copy(isActive = it.id == serverId) }
    }

    override suspend fun removeServer(serverId: String) {
        failWith?.let { throw it }
        state.value = state.value.filterNot { it.id == serverId }
    }

    override suspend fun loginUrl(serverId: String): String {
        failOnce?.let {
            failOnce = null
            throw it
        }
        failWith?.let { throw it }
        return "https://beam.test/v1/auth/login?redirect=/"
    }

    override suspend fun completeLogin(
        serverId: String,
        sessionCookie: String,
    ): UserSummary {
        capturedCookie = sessionCookie
        failWith?.let { throw it }
        val user = Fixtures.user()
        state.value =
            state.value.map {
                if (it.id == serverId) it.copy(state = SessionState.Authenticated(user)) else it
            }
        return user
    }

    override suspend fun logout(serverId: String) {
        state.value =
            state.value.map {
                if (it.id == serverId) it.copy(state = SessionState.LoggedOut) else it
            }
    }

    /** Fingerprints accepted, per server. */
    public val trusted: MutableMap<String, MutableList<String>> = mutableMapOf()

    /**
     * Set to make the next call throw once, then clear itself.
     *
     * Distinct from [failWith], which fails everything: a trust prompt is
     * driven by a failure that must *stop* failing once the certificate is
     * accepted, and a permanent failure cannot express that.
     */
    public var failOnce: BeamException? = null

    override suspend fun trustCertificate(
        serverId: String,
        fingerprint: String,
    ) {
        failWith?.let { throw it }
        trusted.getOrPut(serverId) { mutableListOf() } += fingerprint
    }

    override suspend fun forgetCertificates(serverId: String) {
        failWith?.let { throw it }
        trusted.remove(serverId)
    }

    override suspend fun trustedCertificates(serverId: String): List<String> {
        failWith?.let { throw it }
        return trusted[serverId].orEmpty()
    }

    override suspend fun sessionState(serverId: String): SessionState =
        state.value.firstOrNull { it.id == serverId }?.state ?: SessionState.LoggedOut

    override suspend fun activeServer(): ServerSummary? = state.value.firstOrNull { it.isActive }
}

/** Signed-in devices, held in memory. */
public class FakeSessionRepository : SessionRepository {
    /** Set to make every call throw. */
    public var failWith: BeamException? = null

    /**
     * The devices returned by [sessions].
     *
     * Populated by default, matching every other fake here: a test asserting
     * on an empty list should say so explicitly rather than depending on the
     * fake happening to start empty.
     */
    public var deviceSessions: MutableList<DeviceSession> =
        mutableListOf(Fixtures.deviceSession())

    /** Whether [logoutEverywhere] was called. */
    public var loggedOutEverywhere: Boolean = false

    override suspend fun sessions(): List<DeviceSession> {
        failWith?.let { throw it }
        return deviceSessions
    }

    override suspend fun revoke(sessionId: String) {
        failWith?.let { throw it }
        deviceSessions.removeAll { it.id == sessionId }
    }

    override suspend fun logoutEverywhere() {
        failWith?.let { throw it }
        loggedOutEverywhere = true
    }
}

/** An administrative surface that answers from memory. */
public class FakeAdminRepository : AdminRepository {
    /** Set to make every call throw, typically `Forbidden`. */
    public var failWith: BeamException? = null

    /** The dashboard snapshot. */
    public var statusValue: AdminStatus = Fixtures.adminStatus()

    /**
     * Accounts, held as mutable state rather than a fixed page.
     *
     * Stateful because the interesting assertion is that blocking an account
     * is *reflected* afterwards -- a fake that only records the call cannot
     * tell a working implementation from one that drops the write.
     */
    public var userList: MutableList<AdminUser> =
        mutableListOf(
            Fixtures.adminUser(),
            Fixtures.adminUser(id = "user-2", displayName = "Grace Hopper", isAdmin = true),
        )

    /**
     * Set to make [users] alone fail.
     *
     * Distinct from [failWith] because the user list is one section of the
     * dashboard: it failing must be distinguishable from the whole screen
     * failing.
     */
    public var usersFailWith: BeamException? = null

    /** Log lines returned by [logs]. */
    public var logEntries: List<AdminLogEntry> = emptyList()

    /** Events returned by [events]. */
    public var eventEntries: List<AdminEvent> = emptyList()

    /** Every (userId, disabled) pair passed to [setUserDisabled]. */
    public val disableCalls: MutableList<Pair<String, Boolean>> = mutableListOf()

    /** Every library id passed to [scanLibrary]. */
    public val scanCalls: MutableList<String> = mutableListOf()

    override suspend fun status(): AdminStatus {
        failWith?.let { throw it }
        return statusValue
    }

    override suspend fun health(): ServerHealth {
        failWith?.let { throw it }
        return ServerHealth("ok", "0.1.0", 86_400uL, "ok")
    }

    override suspend fun users(
        limit: UInt?,
        offset: UInt?,
    ): AdminUserPage {
        usersFailWith?.let { throw it }
        failWith?.let { throw it }
        return AdminUserPage(userList, userList.size.toULong())
    }

    override suspend fun setUserDisabled(
        userId: String,
        disabled: Boolean,
    ) {
        failWith?.let { throw it }
        disableCalls += userId to disabled
        val index = userList.indexOfFirst { it.id == userId }
        if (index >= 0) {
            userList[index] = userList[index].copy(disabled = disabled)
        }
    }

    override suspend fun logs(
        limit: UInt?,
        offset: UInt?,
    ): List<AdminLogEntry> {
        failWith?.let { throw it }
        return logEntries
    }

    override suspend fun logCount(): ULong {
        failWith?.let { throw it }
        return logEntries.size.toULong()
    }

    override suspend fun events(limit: UInt?): List<AdminEvent> {
        failWith?.let { throw it }
        return eventEntries
    }

    override suspend fun createLibrary(
        name: String,
        rootPath: String,
    ): LibrarySummary {
        failWith?.let { throw it }
        return Fixtures.library(name = name)
    }

    override suspend fun deleteLibrary(libraryId: String) {
        failWith?.let { throw it }
    }

    override suspend fun scanLibrary(libraryId: String): UInt {
        failWith?.let { throw it }
        scanCalls += libraryId
        return 7u
    }

    override suspend fun refreshMetadata(mediaId: String) {
        failWith?.let { throw it }
    }
}

/** One position report captured by [FakePlaybackRepository]. */
public data class ProgressReport(
    /** The file the position belongs to. */
    val fileId: String,
    /** Where the viewer had got to, in seconds. */
    val positionSecs: Double,
    /** The title's length, where the player knew it. */
    val durationSecs: Double?,
    /** Whether the core was asked to bypass its throttle. */
    val force: Boolean,
)

/**
 * Preferences held in memory.
 *
 * A real [dev.beam.android.core.ffi.preferences.PreferencesRepository] is
 * DataStore-backed, which means a file, a coroutine scope and a real
 * filesystem -- none of which a view-model test should need.
 */
public class FakePreferencesRepository(
    initial: UserPreferences = UserPreferences(),
) : PreferencesRepository {
    private val state = MutableStateFlow(initial)

    override val preferences: Flow<UserPreferences> = state

    /** The current value, for assertions. */
    public val current: UserPreferences get() = state.value

    /** How many times [update] was called. */
    public var updateCount: Int = 0
        private set

    override suspend fun update(transform: (UserPreferences) -> UserPreferences) {
        updateCount++
        state.value = transform(state.value)
    }
}

File: beam-apple/Packages/BeamKit/Sources/BeamTesting/FakeRepositories.swift
import BeamCore
import BeamFFI
import BeamModel
import Foundation
import os

/// Stateful fakes, one per repository seam.
///
/// Stateful rather than mocked, deliberately: a screen that adds a server and
/// then lists servers should see it, and only a fake that actually holds state
/// can prove that. A canned-expectation mock would let a screen "pass" while
/// never reading back what it wrote. Mirrors `core/testing/FakeRepositories.kt`.
///
/// These are scaffolding and never the subject of a test. A test that asserts
/// `FakeCatalogRepository` returns what it was configured with proves nothing
/// about the app.

/// State that can be told to fail.
///
/// Every fake carries the same knob, and honouring it is written once here so
/// no fake can silently forget to -- which would leave that repository's error
/// branches untestable while looking as if they were covered.
private protocol FailureHolding {
    var failure: BeamFailure? { get }
}

extension FailureHolding {
    /// Throw the configured failure, if there is one.
    func check() throws {
        if let failure { throw failure }
    }
}

/// A `ServerRepository` that remembers what it was told.
public final class FakeServerRepository: ServerRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: State())

    private struct State: FailureHolding {
        var servers: [ServerSummary] = []
        var trusted: [String: Set<String>] = [:]
        var failure: BeamFailure?
        var completedCookies: [String: String] = [:]
    }

    /// An empty registry.
    public init(servers: [ServerSummary] = []) {
        state.withLock { $0.servers = servers }
    }

    /// Make every call fail with `failure`, or `nil` to stop failing.
    ///
    /// The knob exists so the error branches -- an unreachable server, an
    /// expired session, a locked keystore -- are reachable from a test rather
    /// than only from a broken network (NFR-205).
    public func fail(with failure: BeamFailure?) {
        state.withLock { $0.failure = failure }
    }

    /// The cookie `completeLogin` was handed, if any.
    public func capturedCookie(serverId: String) -> String? {
        state.withLock { $0.completedCookies[serverId] }
    }

    public func restore() async throws -> [ServerSummary] {
        try state.withLock {
            try $0.check(); return $0.servers
        }
    }

    public func listServers() throws -> [ServerSummary] {
        try state.withLock {
            try $0.check(); return $0.servers
        }
    }

    public func addServer(baseURL: String, displayName: String?) async throws -> ServerSummary {
        try state.withLock { state in
            try state.check()
            if let existing = state.servers.first(where: { $0.baseUrl == baseURL }) {
                return existing
            }
            let summary = ServerSummary(
                id: "server-\(state.servers.count + 1)",
                displayName: displayName ?? baseURL,
                baseUrl: baseURL,
                state: .loggedOut,
                isActive: state.servers.isEmpty
            )
            state.servers.append(summary)
            return summary
        }
    }

    public func selectServer(id serverId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.servers = state.servers.map { server in
                ServerSummary(
                    id: server.id,
                    displayName: server.displayName,
                    baseUrl: server.baseUrl,
                    state: server.state,
                    isActive: server.id == serverId
                )
            }
        }
    }

    public func removeServer(id serverId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.servers.removeAll { $0.id == serverId }
            state.trusted.removeValue(forKey: serverId)
        }
    }

    public func loginURL(serverId: String) throws -> String {
        try state.withLock { state in
            try state.check()
            guard let server = state.servers.first(where: { $0.id == serverId }) else {
                throw BeamFailure(message: "unknown server", requiresSignIn: true)
            }
            return "\(server.baseUrl)/v1/auth/login?redirect=/"
        }
    }

    public func completeLogin(serverId: String, sessionCookie: String) async throws -> UserSummary {
        try state.withLock { state in
            try state.check()
            state.completedCookies[serverId] = sessionCookie
            let user = UserSummary(
                id: "user-1",
                displayName: "Viewer",
                email: "viewer@example.invalid",
                isAdmin: false,
                avatarUrl: nil
            )
            state.servers = state.servers.map { server in
                guard server.id == serverId else { return server }
                return ServerSummary(
                    id: server.id,
                    displayName: server.displayName,
                    baseUrl: server.baseUrl,
                    state: .authenticated(user: user),
                    isActive: server.isActive
                )
            }
            return user
        }
    }

    public func logout(serverId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.servers = state.servers.map { server in
                guard server.id == serverId else { return server }
                return ServerSummary(
                    id: server.id,
                    displayName: server.displayName,
                    baseUrl: server.baseUrl,
                    state: .loggedOut,
                    isActive: server.isActive
                )
            }
        }
    }

    public func sessionState(serverId: String) throws -> SessionState {
        try state.withLock { state in
            try state.check()
            return state.servers.first { $0.id == serverId }?.state ?? .loggedOut
        }
    }

    public func trustCertificate(serverId: String, fingerprint: String) async throws {
        try state.withLock { state in
            try state.check()
            state.trusted[serverId, default: []].insert(fingerprint)
        }
    }

    public func forgetCertificates(serverId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.trusted[serverId] = []
        }
    }

    public func trustedCertificates(serverId: String) throws -> [String] {
        try state.withLock { state in
            try state.check()
            return Array(state.trusted[serverId] ?? [])
        }
    }
}

/// A `CatalogRepository` over an in-memory catalogue.
public final class FakeCatalogRepository: CatalogRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: State())

    private struct State: FailureHolding {
        var items: [MediaSummary] = []
        var details: [String: MediaDetail] = [:]
        var libraries: [LibrarySummary] = []
        var files: [String: [LibraryFileSummary]] = [:]
        var genres: [String] = []
        var upNext: EpisodeSummary?
        var failure: BeamFailure?
        var lastQuery: BrowseQuery?
    }

    /// A catalogue holding `items`.
    public init(items: [MediaSummary] = [], libraries: [LibrarySummary] = []) {
        state.withLock {
            $0.items = items
            $0.libraries = libraries
            $0.genres = Array(Set(items.flatMap(\.genres))).sorted()
        }
    }

    /// Make every call fail.
    public func fail(with failure: BeamFailure?) {
        state.withLock { $0.failure = failure }
    }

    /// Provide the detail returned for `mediaId`.
    public func setDetail(_ detail: MediaDetail, for mediaId: String) {
        state.withLock { $0.details[mediaId] = detail }
    }

    /// Provide the next episode `upNext` should resolve to.
    public func setUpNext(_ episode: EpisodeSummary?) {
        state.withLock { $0.upNext = episode }
    }

    /// The query the last `browse` was given.
    ///
    /// Lets a test assert what the screen actually asked for -- the debounce
    /// interval, the page size, the filters -- rather than only what came back.
    public func lastQuery() -> BrowseQuery? {
        state.withLock { $0.lastQuery }
    }

    public func browse(query: BrowseQuery) async throws -> MediaPage {
        try state.withLock { state in
            try state.check()
            state.lastQuery = query
            var items = state.items
            if let search = query.query, !search.isEmpty {
                items = items.filter { $0.title.localizedCaseInsensitiveContains(search) }
            }
            if let genre = query.genre {
                items = items.filter { $0.genres.contains(genre) }
            }
            return Fixtures.page(items)
        }
    }

    public func detail(mediaId: String) async throws -> MediaDetail {
        try state.withLock { state in
            try state.check()
            guard let detail = state.details[mediaId] else {
                throw BeamFailure(message: "not found")
            }
            return detail
        }
    }

    public func genres() async throws -> [String] {
        try state.withLock {
            try $0.check(); return $0.genres
        }
    }

    public func upNext(showId: String, currentEpisodeId: String) async throws -> EpisodeSummary? {
        try state.withLock {
            try $0.check(); return $0.upNext
        }
    }

    public func libraries() async throws -> [LibrarySummary] {
        try state.withLock {
            try $0.check(); return $0.libraries
        }
    }

    public func library(id libraryId: String) async throws -> LibrarySummary {
        try state.withLock { state in
            try state.check()
            guard let library = state.libraries.first(where: { $0.id == libraryId }) else {
                throw BeamFailure(message: "not found")
            }
            return library
        }
    }

    public func libraryFiles(libraryId: String) async throws -> [LibraryFileSummary] {
        try state.withLock {
            try $0.check(); return $0.files[libraryId] ?? []
        }
    }
}

/// A `PlaybackRepository` that records what it was told.
public final class FakePlaybackRepository: PlaybackRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: State())

    private struct State: FailureHolding {
        var sources: [String: [MediaSourceView]] = [:]
        var selection: SourceSelection?
        var continueWatching: [ContinueWatchingEntry] = []
        var history: [HistoryEntry] = []
        var reports: [(fileId: String, position: Double, forced: Bool)] = []
        var profile: DeviceProfile?
        var failure: BeamFailure?
    }

    /// A repository with nothing in it.
    public init(
        continueWatching: [ContinueWatchingEntry] = [],
        history: [HistoryEntry] = []
    ) {
        state.withLock {
            $0.continueWatching = continueWatching
            $0.history = history
        }
    }

    /// Make every call fail.
    public func fail(with failure: BeamFailure?) {
        state.withLock { $0.failure = failure }
    }

    /// Provide the sources returned for `mediaId`.
    public func setSources(_ sources: [MediaSourceView], for mediaId: String) {
        state.withLock { $0.sources[mediaId] = sources }
    }

    /// Provide the selection `selectSource` should return.
    public func setSelection(_ selection: SourceSelection?) {
        state.withLock { $0.selection = selection }
    }

    /// Every progress report made, in order.
    public func reports() -> [(fileId: String, position: Double, forced: Bool)] {
        state.withLock { $0.reports }
    }

    /// The profile the app pushed at launch.
    public func deviceProfile() -> DeviceProfile? {
        state.withLock { $0.profile }
    }

    public func setDeviceProfile(_ profile: DeviceProfile) {
        state.withLock { $0.profile = profile }
    }

    public func sources(mediaId: String) async throws -> [MediaSourceView] {
        try state.withLock {
            try $0.check(); return $0.sources[mediaId] ?? []
        }
    }

    public func selectSource(mediaId: String, policy: QualityPolicy) async throws -> SourceSelection
    {
        try state.withLock { state in
            try state.check()
            guard let selection = state.selection else {
                throw BeamFailure(message: "no playable source")
            }
            return selection
        }
    }

    public func playbackConfig(fileId: String) throws -> PlaybackHttpConfig {
        try state.withLock {
            try $0.check(); return Fixtures.playbackConfig(fileId: fileId)
        }
    }

    public func serverHTTPConfig() throws -> ServerHttpConfig {
        try state.withLock { state in
            try state.check()
            return ServerHttpConfig(
                baseUrl: "https://beam.invalid",
                headers: ["Cookie": "beam_session=opaque"],
                trustedFingerprints: [],
                host: "beam.invalid"
            )
        }
    }

    public func reportProgress(
        fileId: String,
        positionSeconds: Double,
        durationSeconds: Double?,
        force: Bool
    ) async throws -> ProgressOutcome {
        try state.withLock { state in
            try state.check()
            state.reports.append((fileId, positionSeconds, force))
            return .sent(positionSecs: positionSeconds)
        }
    }

    public func flushProgress() async throws -> UInt32 {
        try state.withLock {
            try $0.check(); return 0
        }
    }

    public func pendingProgressCount() async throws -> UInt32 {
        try state.withLock {
            try $0.check(); return 0
        }
    }

    public func continueWatching(limit: UInt32?) async throws -> [ContinueWatchingEntry] {
        try state.withLock {
            try $0.check(); return $0.continueWatching
        }
    }

    public func history(limit: UInt32?, offset: UInt32?) async throws -> HistoryPage {
        try state.withLock { state in
            try state.check()
            let start = Int(offset ?? 0)
            let end = min(state.history.count, start + Int(limit ?? 50))
            let slice = start < end ? Array(state.history[start..<end]) : []
            return HistoryPage(items: slice, total: UInt64(state.history.count))
        }
    }
}

/// A `SessionRepository` over an in-memory list.
public final class FakeSessionRepository: SessionRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: State())

    private struct State: FailureHolding {
        var sessions: [DeviceSession] = []
        var failure: BeamFailure?
    }

    /// A repository holding `sessions`.
    public init(sessions: [DeviceSession] = []) {
        state.withLock { $0.sessions = sessions }
    }

    /// Make every call fail.
    public func fail(with failure: BeamFailure?) {
        state.withLock { $0.failure = failure }
    }

    public func sessions() async throws -> [DeviceSession] {
        try state.withLock {
            try $0.check(); return $0.sessions
        }
    }

    public func revoke(sessionId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.sessions.removeAll { $0.id == sessionId }
        }
    }

    public func logoutEverywhere() async throws {
        try state.withLock { state in
            try state.check()
            state.sessions = []
        }
    }
}

/// An `AdminRepository` over in-memory state.
public final class FakeAdminRepository: AdminRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: State())

    private struct State: FailureHolding {
        var status: AdminStatus?
        var users: [AdminUser] = []
        var logs: [AdminLogEntry] = []
        var events: [AdminEvent] = []
        var libraries: [LibrarySummary] = []
        var scanned: [String] = []
        var failure: BeamFailure?
    }

    /// A repository with nothing in it.
    public init(status: AdminStatus? = nil, users: [AdminUser] = []) {
        state.withLock {
            $0.status = status
            $0.users = users
        }
    }

    /// Make every call fail. A `.isForbidden` failure is how a
    /// non-administrator is simulated, matching the server's own 403.
    public func fail(with failure: BeamFailure?) {
        state.withLock { $0.failure = failure }
    }

    /// Which libraries were scanned, in order.
    public func scannedLibraries() -> [String] {
        state.withLock { $0.scanned }
    }

    public func status() async throws -> AdminStatus {
        try state.withLock { state in
            try state.check()
            guard let status = state.status else { throw BeamFailure(message: "no status") }
            return status
        }
    }

    public func users(limit: UInt32?, offset: UInt32?) async throws -> AdminUserPage {
        try state.withLock { state in
            try state.check()
            return AdminUserPage(items: state.users, total: UInt64(state.users.count))
        }
    }

    public func setUserDisabled(userId: String, disabled: Bool) async throws {
        try state.withLock { (state: inout State) in
            try state.check()
            state.users = state.users.map { user in
                guard user.id == userId else { return user }
                return AdminUser(
                    id: user.id,
                    displayName: user.displayName,
                    email: user.email,
                    avatarUrl: user.avatarUrl,
                    isAdmin: user.isAdmin,
                    disabled: disabled,
                    createdAtUnix: user.createdAtUnix
                )
            }
        }
    }

    public func logs(limit: UInt32?, offset: UInt32?) async throws -> [AdminLogEntry] {
        try state.withLock {
            try $0.check(); return $0.logs
        }
    }

    public func logCount() async throws -> UInt64 {
        try state.withLock {
            try $0.check(); return UInt64($0.logs.count)
        }
    }

    public func events(limit: UInt32?) async throws -> [AdminEvent] {
        try state.withLock {
            try $0.check(); return $0.events
        }
    }

    public func createLibrary(name: String, rootPath: String) async throws -> LibrarySummary {
        try state.withLock { state in
            try state.check()
            let library = Fixtures.library(
                id: "library-\(state.libraries.count + 1)",
                name: name,
                size: 0
            )
            state.libraries.append(library)
            return library
        }
    }

    public func deleteLibrary(id libraryId: String) async throws {
        try state.withLock { state in
            try state.check()
            state.libraries.removeAll { $0.id == libraryId }
        }
    }

    public func scanLibrary(id libraryId: String) async throws -> UInt32 {
        try state.withLock { state in
            try state.check()
            state.scanned.append(libraryId)
            return 7
        }
    }

    public func refreshMetadata(mediaId: String) async throws {
        try state.withLock { try $0.check() }
    }

    public func health() async throws -> ServerHealth {
        try state.withLock { state in
            try state.check()
            return ServerHealth(
                status: "ok",
                version: "0.1.0",
                uptimeSecs: 3600,
                database: "ok"
            )
        }
    }
}

/// A `PreferencesRepository` that holds preferences in memory.
public final class FakePreferencesRepository: PreferencesRepository, @unchecked Sendable {
    private let state = OSAllocatedUnfairLock(initialState: UserPreferences.default)

    /// A repository starting from `preferences`.
    public init(_ preferences: UserPreferences = .default) {
        state.withLock { $0 = preferences }
    }

    public func load() -> UserPreferences {
        state.withLock { $0 }
    }

    public func save(_ preferences: UserPreferences) {
        state.withLock { $0 = preferences }
    }
}
